File: 2018/day/9/2.cpp
#include <list>
#include <vector>
#include <iostream>

int main(int argc, char *argv[]) {
  auto players = std::atoi(argv[1]);
  auto marbles = std::atoi(argv[2]);
  std::list<int64_t> circle;
  std::vector<int64_t> score;
  score.resize(players);
  circle.push_back(0);
  auto it = circle.begin();
  auto player = 1;
  for (int marble = 1; marble < marbles + 1; ++marble) {
    if ((marble % 23) == 0) {
      for (int x = 0; x < 7; ++x) {
        if (it == circle.begin()) {
          it = circle.end();
        }
        --it;
      }
      auto value = *it;
      it = circle.erase(it);
      score[player] += (value + marble);
    } else {
      for (int x = 0; x < 2; ++x) {
        ++it;
        if (it == circle.end()) {
          it = circle.begin();
        }
      }
      it = circle.insert(it, marble);
    }
    player = (player + 1) % players;
    if (marble % 10000 == 0) {
      std::cout << marble << " " << (100.0 * marble) / marbles << "%\n";
    }
  }
  int64_t s = 0;
  for (auto sc : score) {
    if (sc > s) {
      s = sc;
    }
  }
  std::cout << s;
}

File: 2018/day/9/2.rs
extern crate linked_list;

use linked_list::LinkedList;
use std::collections::HashMap;
use std::env;

fn solve(players: i64, marbles: i64) -> i64 {
    let mut circle = LinkedList::new();
    circle.push_front(0);
    let mut cursor = circle.cursor();
    let mut score = HashMap::new();
    for marble in 1..(marbles+1) {
        if marble % 23 == 0 {
            for _ in 0..7 {
                if let None = cursor.prev() {
                    cursor.prev();
                }
            }
            let removed = cursor.remove().unwrap();
            *score.entry(marble % players).or_insert(0) += removed + marble;
        } else {
            for _ in 0..2 {
                if let None = cursor.next() {
                    cursor.next();
                }
            }
            cursor.insert(marble);
        }
    }
    return *score.iter().map(|a| a.1).max().unwrap();
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let players = args[1].parse::<i64>().unwrap();
    let marbles = args[2].parse::<i64>().unwrap();

    let result = solve(players, marbles);
    println!("{}", result);
}
